### PR TITLE
Open API

### PR DESCRIPTION
**Pull request by xinoxapps**
Added support for the OpenAPI action at the server side.

### CODE REVIEW

**Comment by xinoxapps** [comment]
I signed it!

**Comment by loosebazooka** [comment]
Thanks @xinoxapps I'm not sure if we intended this plugin to cover the open-api generation. However, I'll take a look and see if we can expand the scope of this plugin.

**Comment by patflynn** [comment]
@xinoxapps and thank you for the PR! We appreciate it.

File: src/main/java/com/google/cloud/tools/gradle/endpoints/framework/server/task/GenerateOpenApiDocsTask.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    List<String> params = new ArrayList<>(Arrays.asList(
+        GetOpenApiDocAction.NAME,
+        "-o", openApiDocDir.getPath() + "/openapi.json",
+        "-cp", classpath,
+        "-w", webAppDir.getPath()));
+    params.addAll(getServiceClasses());
+
+    new EndpointsTool().execute(params.toArray(new String[params.size()]));
+  }
+}
```
Comment (duchuyctlk) [inline src/main/java/com/google/cloud/tools/gradle/endpoints/framework/server/task/GenerateOpenApiDocsTask.java]: Please add a new line at end of file 😄

**Comment by loosebazooka** [comment]
@xinoxapps Can you explain how you intend to use this? We don't really want to include the open-api related features in here. There is generic open-api tooling for gradle that makes sense to use. This is merely a way to bring our previous endpoints functionality up to date, however I believe a compelling use case might sway us.

**Comment by xinoxapps** [comment]
Hi,

Thank you for your positive feedback. We are also in the progress of migrating our existing projects to the new endpoints framework. 

We are currently missing a convenient way to generate openapi docs for the Google Console Dashboard. Which applies to the following command:

`gcloud service-management deploy openapi.json`

In the Google Docs, there is a [section](https://cloud.google.com/endpoints/docs/frameworks/java/about-cloud-endpoints-frameworks) under tooling about endpoints.cmd. I think this plugin fulfills the role as convenient gradle wrapper. 

> The endpoints command line tool (for Linux), or endpoints.cmd (for Windows) that you can use to generate client libraries, discovery documents, and OpenApi specification configuration file required for Cloud Endpoints

Another compelling use case is to support the following [typescript client library](https://www.npmjs.com/package/openapi-client)

Kind regards, Wendel

File: build.gradle [hunk 2]
```diff
@@ -17,7 +17,7 @@ sourceCompatibility = JavaVersion.VERSION_1_7
 targetCompatibility = JavaVersion.VERSION_1_7
 
 group = 'com.google.cloud.tools'
-version = '1.0.0-SNAPSHOT'
+version = '1.0.1-SNAPSHOT'
```
Comment (loosebazooka) [inline build.gradle]: Please change this back to 1.0.0

File: src/main/java/com/google/cloud/tools/gradle/endpoints/framework/client/EndpointsClientPlugin.java [hunk 3]
```diff
@@ -136,6 +133,10 @@ public void execute(Project project) {
     });
   }
 
+
```
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/gradle/endpoints/framework/client/EndpointsClientPlugin.java]: Undo all these newlines.

File: src/main/java/com/google/cloud/tools/gradle/endpoints/framework/client/EndpointsClientExtension.java [hunk 4]
```diff
@@ -30,16 +30,20 @@
   private File genSrcDir;
   private final File clientLibDir;
   private final File genDiscoveryDocsDir;
+  private final File genOpenApiDocsDir;
```
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/gradle/endpoints/framework/client/EndpointsClientExtension.java]: It does not appear likely that we need to modify the client plugin. The client plugin is designed for generating client libraries for client consumers of endpoints v1.1.

For OpenApi specs, I believe you're going to delegate to another tool. Are you using something to do that? And how are you using it?

File: src/main/java/com/google/cloud/tools/gradle/endpoints/framework/server/task/GenerateOpenApiDocsTask.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
+/*
+ *  Copyright (c) 2016 Google Inc. All Right Reserved.
```
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/gradle/endpoints/framework/server/task/GenerateOpenApiDocsTask.java]: 2017

File: src/test/java/com/google/cloud/tools/gradle/endpoints/framework/ProjectTests.java [hunk 6]
```diff
@@ -59,14 +58,17 @@ public void testClientBuilds() throws IOException, URISyntaxException {
...
   public void testServer() throws IOException, URISyntaxException {
-    FileUtils.copyDirectory(
-        new File(getClass().getClassLoader().getResource("projects/server").toURI()),
-        testProjectDir.getRoot());
-    BuildResult buildResult = GradleRunner.create()
-        .withProjectDir(testProjectDir.getRoot())
-        .withPluginClasspath()
-        .withArguments("endpointsClientLibs", "endpointsDiscoveryDocs")
-        .build();
+
```
Comment (loosebazooka) [inline src/test/java/com/google/cloud/tools/gradle/endpoints/framework/ProjectTests.java]: Looks like you reformatted this with 4 spaces? Undo that.

File: src/main/java/com/google/cloud/tools/gradle/endpoints/framework/server/EndpointsServerPlugin.java [hunk 7]
```diff
@@ -104,6 +108,32 @@ public void execute(Project project) {
         });
   }
 
+
+  private void createGenerateOpenApiDocsTask() {
+    project.getTasks().create(GENERATE_OPENAPI_DOC_TASK, GenerateOpenApiDocsTask.class,
+            new Action<GenerateOpenApiDocsTask>() {
```
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/gradle/endpoints/framework/server/EndpointsServerPlugin.java]: spacing again, use +4 for line continuations.

**Review by loosebazooka — CHANGES_REQUESTED** [review]

**Comment by loosebazooka** [comment]
@xinoxapps okay, looks like we're going to try and merge this in. Please update based on the review, and thanks again for the contribution.

**Comment by xinoxapps** [comment]
Requested changes applied

File: src/main/java/com/google/cloud/tools/gradle/endpoints/framework/server/task/GenerateDiscoveryDocsTask.java [hunk 8]
```diff
@@ -17,10 +17,7 @@
 
 import com.google.api.server.spi.tools.EndpointsTool;
 import com.google.api.server.spi.tools.GetDiscoveryDocAction;
-import java.io.File;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.List;
+
```
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/gradle/endpoints/framework/server/task/GenerateDiscoveryDocsTask.java:20]: Can you remove the changes you made to this file?

File: src/test/resources/projects/client/src/endpoints/testApi-v1-rest.openapi [hunk 9]
```diff
@@ -0,0 +1,109 @@
+{
```
Comment (loosebazooka) [inline src/test/resources/projects/client/src/endpoints/testApi-v1-rest.openapi:1]: I think we can remove this file.

File: src/main/java/com/google/cloud/tools/gradle/endpoints/framework/server/EndpointsServerPlugin.java [hunk 10]
```diff
@@ -75,59 +79,83 @@ public void execute(Project project) {
...
-                genClientLibs.setClassesDir(classesDir);
-                genClientLibs.setClientLibDir(extension.getClientLibDir());
-                genClientLibs.setServiceClasses(extension.getServiceClasses());
-                genClientLibs.setWebAppDir(
-                    project.getConvention().getPlugin(WarPluginConvention.class).getWebAppDir());
-              }
-            });
-          }
-        });
+      new Action<GenerateClientLibsTask>() {
```
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/gradle/endpoints/framework/server/EndpointsServerPlugin.java]: please undo the spacing changes here, they were correct before (offset by 2x for line continuation)

File: src/main/java/com/google/cloud/tools/gradle/endpoints/framework/server/EndpointsServerPlugin.java [hunk 11]
```diff
@@ -75,59 +79,83 @@ public void execute(Project project) {
...
-                genDiscoveryDocs.setClassesDir(classesDir);
-                genDiscoveryDocs.setDiscoveryDocDir(extension.getDiscoveryDocDir());
-                genDiscoveryDocs.setServiceClasses(extension.getServiceClasses());
-                genDiscoveryDocs.setWebAppDir(
-                    project.getConvention().getPlugin(WarPluginConvention.class).getWebAppDir());
-              }
-            });
-          }
-        });
+      new Action<GenerateDiscoveryDocsTask>() {
```
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/gradle/endpoints/framework/server/EndpointsServerPlugin.java]: please undo the spacing changes here, they were correct before (offset by 2x for line continuation)

File: src/main/java/com/google/cloud/tools/gradle/endpoints/framework/server/EndpointsServerPlugin.java [hunk 12]
```diff
@@ -75,59 +79,83 @@ public void execute(Project project) {
...
+                  project.getConvention().getPlugin(WarPluginConvention.class).getWebAppDir());
+            }
+          });
+        }
+      });
+  }
+
+  private void createGenerateOpenApiDocsTask() {
+    project.getTasks().create(GENERATE_OPENAPI_DOC_TASK, GenerateOpenApiDocsTask.class,
+      new Action<GenerateOpenApiDocsTask>() {
```
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/gradle/endpoints/framework/server/EndpointsServerPlugin.java:112]: I think you need to correct the spacing here +2 spaces from 112 to 133?

File: README.md [hunk 13]
```diff
@@ -33,6 +33,7 @@ apply plugin: 'com.google.cloud.tools.endpoints-framework-server'
 The plugin exposes the following server side goals
 * `endpointsClientLibs` - generate client libraries
 * `endpointsDiscoveryDocs` - generate discovery documents
+* `endpointsOpenApiDocs` - generate openapi documents
```
Comment (loosebazooka) [inline README.md:36]: Docs will not change until a release. I'll make this update when we do the release.

**Review by loosebazooka — CHANGES_REQUESTED** [review]

**Comment by loosebazooka** [comment]
The plugin now has checkstlye and google-java-format built in.

You should be able to run `./gradlew googleJavaFormat` to auto format most files.

**Comment by xinoxapps** [comment]
I did a merge with the style branch but ended with several conflicts.

gradle googleJavaFormat has been applied to the srces.

**Comment by loosebazooka** [comment]
@xinoxapps I cleaned up and added to this PR, in #40 , but it needs your approval, since it contains a commit from you.

**Pull request closed without merging**